src/librustc_typeck/astconv.rs

```diff
@@ -1439,6 +1461,134 @@ impl<'o, 'gcx: 'tcx, 'tcx> dyn AstConv<'gcx, 'tcx> + 'o {
+        // Luckily, we can (at least for now) deduce the intermediate steps
+        // just from the end-point.
+        //
+        // There are basically five cases to consider:
+        //
+        // 1. Reference to a constructor of a struct:
+        //
+        //        struct Foo<T>(...)
+        //
+        //    In this case, the parameters are declared in the type space.
+        //
+        // 2. Reference to a constructor of an enum variant:
+        //
+        //        enum E<T> { Foo(...) }
+        //
+        //    In this case, the parameters are defined in the type space,
+        //    but may be specified either on the type or the variant.
+        //
+        // 3. Reference to a fn item or a free constant:
+        //
+        //        fn foo<T>() { }
+        //
+        //    In this case, the path will again always have the form
+        //    `a::b::foo::<T>` where only the final segment should have
+        //    type parameters. However, in this case, those parameters are
+        //    declared on a value, and hence are in the `FnSpace`.
+        //
+        // 4. Reference to a method or an associated constant:
+        //
+        //        impl<A> SomeStruct<A> {
+        //            fn foo<B>(...)
+        //        }
+        //
+        //    Here we can have a path like
+        //    `a::b::SomeStruct::<A>::foo::<B>`, in which case parameters
+        //    may appear in two places. The penultimate segment,
+        //    `SomeStruct::<A>`, contains parameters in TypeSpace, and the
+        //    final segment, `foo::<B>` contains parameters in fn space.
+        //
+        // 5. Reference to a local variable
+        //
+        //    Local variables can't have any type parameters.
+        //
+        // The first step then is to categorize the segments appropriately.
+
+        let tcx = self.tcx();
+
+        assert!(!segments.is_empty());
+        let last = segments.len() - 1;
+
+        let mut path_segs = vec![];
+
+        match def {
+            // Case 1. Reference to a struct constructor.
+            Def::StructCtor(def_id, ..) |
+            Def::SelfCtor(.., def_id) => {
+                // Everything but the final segment should have no
+                // parameters at all.
+                let generics = tcx.generics_of(def_id);
+                // Variant and struct constructors use the
+                // generics of their parent type definition.
+                let generics_def_id = generics.parent.unwrap_or(def_id);
+                path_segs.push(PathSeg(generics_def_id, last));
+            }
+
+            // Case 2. Reference to a variant constructor.
+            Def::Variant(def_id) |
+            Def::VariantCtor(def_id, ..) => {
+                let adt_def = self_ty.and_then(|t| t.ty_adt_def());
+                let (generics_def_id, index) = if let Some(adt_def) = adt_def {
+                    debug_assert!(adt_def.is_enum());
+                    (adt_def.did, last)
```

Comment:
Well, there's a difference between this and what `Def::Method` below does (after which we need to model our associated variants).
Equivalent of `path_segs.push(PathSeg(generics.parent.unwrap(), last - 1));` below is missing, so we have the type inference issues.
And equivalent of `path_segs.push(PathSeg(def_id, last));` below shouldn't exist here because variants don't have their own generics.

Reply:
We still need to handle the case the generics are specified on the variant though... how do we accommodate both?

Reply:
`adt_def.is_some()` is true only when the variant is on alias, we don't need to support generics on the variant in this case.

Reply:
Ah, so it's only that branch of the if-statement that's wrong?

Reply:
Literally a change of `last` to `last - 1` I guess you mean? Let me try that out, but I think the problem is for a variant on an alias, the last segment is the *only* segment, so `last - 1` will be an invalid index.

Reply:
Now I'm pretty sure the place you need is in `instantiate_value_path`, right below `def_ids_for_path_segments` where
```rust
            Def::Method(def_id) |
            Def::AssociatedConst(def_id) => {
```
is processed.
`Def::VariantCtor` basically needs to do the equivalent thing - make `user_self_ty` so it's later fed to type inference (`self.at(&self.misc(span), self.param_env).sup(impl_ty, self_ty)`).

Reply:
Aha, that makes sense now! Let me try...